docs/dependency-injection.md

```diff
@@ -0,0 +1,193 @@
+
+The following example shows how to use `@inject`. The service implementing `Services.IDataAccess` is injected into the component's property `DataRepository`. Note how the code is only using the `IDataAccess` abstraction:
+
+```csharp
+@page "/customer-list"
+@using Services
+@inject IDataAccess DataRepository
+
+<ul>
+    @if (Customers != null)
+    {
+        @foreach (var customer in Customers)
+        {
+            <li>@customer.FirstName @customer.LastName</li>
+        }
+    }
+</ul>
+
+@functions {
+    private IReadOnlyList<Customer> Customers;
+
+    protected override async Task OnInitAsync()
+    {
+        // The property DataRepository has received an implemenation
+        // of IDataAccess through dependency injection. We can use
+        // it to get data from our server.
+        Customers = await DataRepository.GetAllCustomersAsync();
+    }
+}
+```
+
+Internally, the generated property (`DataRepository`) is decorated with the `Microsoft.AspNetCore.Blazor.Components.InjectAttribute` attribute. Typically, this attribute isn't used directly. If a base class is required for components and injected properties are also required for the base class, `InjectAttribute` can be manually added:
+
+```csharp
+public class ComponentBase : BlazorComponent
+{
+    // Note that Blazor's dependency injection works even if using the
+    // InjectAttribute in a component's base class.
+    [Inject]
+    protected IDataAccess DataRepository { get; set; }
+    ...
+}
+```
+
+In components derived from the base class, the `@inject` directive isn't required. The `InjectAttribute` of the base class is sufficient:
+
+```csharp
+@page "/demo"
+@inherits ComponentBase
+
+<h1>...</h1>
+...
+```
+
+## Dependency injection in services
+
+Complex services might require additional services. In the prior example, `DataAccess` might require Blazor's default service `HttpClient`. `@inject` or the `InjectAttribute` can't be used in services. *Constructor injection* must be used instead. Required services are added by adding parameters to the service's constructor. When dependency injection creates the service, it recognizes the services it requires in the constructor and provides them accordingly.
+
+The following code sample demonstrates the concept:
+
+```csharp
+public class DataAccess : IDataAccess
+{
+    // Note that the constructor receives an HttpClient via dependency
+    // injection. HttpClient is a default service offered by Blazor.
+    public Repository(HttpClient client)
+    {
+        ...
+    }
+    ...
+}
+```
+
+Note the following prerequisites for constructor injection:
+
+* There must be one constructor whose arguments can all be fulfilled by dependency injection. Note that additional parameters not covered by DI are allowed if default values are specified for them.
+* The applicable constructor must be *public*.
+* There must only be one applicable constructor. In case of an ambiguity, DI throws an exception.
+
+## Service lifetime
```

Comment:
Services are created by the service provider, so they need to be disposed by the service provider. We need to address this as part of https://github.com/aspnet/Blazor/issues/464, but regardless I think this section should be removed as it contradicts that guidance. The content about implementing `IDisposable` on components should be covered in the components doc, not here.